File: src/modules/equipment-asset/services/equipment-asset.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { RETIRE_GRACE_PERIOD_DAYS } from '../../../common/constants/workflow.constants';
import { AccessControlService } from '../../../common/services/access-control.service';
import { AuthenticatedUser } from '../../../common/types/authenticated-user.type';
import { Employee } from '../../employee/entities/employee.entity';
import { EquipmentAsset } from '../entities/equipment-asset.entity';
import { EquipmentAssetStatus } from '../enums/equipment-asset-status.enum';
import {
  AssignEquipmentAssetDto,
  CreateEquipmentAssetDto,
  UpdateEquipmentAssetDto,
  UpdateEquipmentAssetStatusDto,
} from '../dto/equipment-asset.dto';
import { EquipmentAssignment } from '../../equipment-assignment/entities/equipment-assignment.entity';
import { EquipmentAssignmentStatus } from '../../equipment-assignment/enums/equipment-assignment-status.enum';
import { EquipmentModel } from '../../equipment-model/entities/equipment-model.entity';
import { EquipmentRequest } from '../../request/entities/equipment-request.entity';
import { RequestStatus } from '../../request/enums/request-status.enum';
import { NotificationService } from '../../notification/services/notification.service';

@Injectable()
export class EquipmentAssetService {
  constructor(
    @InjectRepository(EquipmentAsset)
    private readonly assetRepository: Repository<EquipmentAsset>,
    @InjectRepository(EquipmentModel)
    private readonly modelRepository: Repository<EquipmentModel>,
    @InjectRepository(EquipmentAssignment)
    private readonly assignmentRepository: Repository<EquipmentAssignment>,
    @InjectRepository(EquipmentRequest)
    private readonly requestRepository: Repository<EquipmentRequest>,
    @InjectRepository(Employee)
    private readonly employeeRepository: Repository<Employee>,
    private readonly notificationService: NotificationService,
    private readonly accessControl: AccessControlService,
  ) {}

  findAll(): Promise<EquipmentAsset[]> {
    return this.assetRepository.find({
      relations: {
        equipmentModel: { category: true },
        assignedEmployee: true,
        retiredBy: true,
      },
      order: { createdAt: 'DESC' },
    });
  }

  async findOne(id: string): Promise<EquipmentAsset> {
    const asset = await this.assetRepository.findOne({
      where: { id },
      relations: {
        equipmentModel: { category: true },
        assignedEmployee: true,
        retiredBy: true,
        assignments: { employee: true },
      },
    });
    if (!asset) {
      throw new NotFoundException(`Equipment asset with id "${id}" not found`);
    }
    return asset;
  }

  async create(dto: CreateEquipmentAssetDto): Promise<EquipmentAsset> {
    const model = await this.modelRepository.findOne({
      where: { id: dto.equipmentModelId },
    });
    if (!model) {
      throw new NotFoundException('Equipment model not found');
    }

    const asset = this.assetRepository.create({
      equipmentModel: model,
      assetTag: dto.assetTag,
      serialNumber: dto.serialNumber,
      status: dto.status ?? EquipmentAssetStatus.AVAILABLE,
      notes: dto.notes,
    });
    return this.assetRepository.save(asset);
  }

  async update(
    id: string,
    dto: UpdateEquipmentAssetDto,
  ): Promise<EquipmentAsset> {
    const asset = await this.findOne(id);
    Object.assign(asset, {
      assetTag: dto.assetTag ?? asset.assetTag,
      serialNumber: dto.serialNumber ?? asset.serialNumber,
      notes: dto.notes ?? asset.notes,
    });
    return this.assetRepository.save(asset);
  }

  async updateStatus(
    id: string,
    dto: UpdateEquipmentAssetStatusDto,
  ): Promise<EquipmentAsset> {
    const asset = await this.findOne(id);
    asset.status = dto.status;
    return this.assetRepository.save(asset);
  }

  async assign(
    id: string,
    dto: AssignEquipmentAssetDto,
    user: AuthenticatedUser,
  ) {
    const asset = await this.findOne(id);
    if (asset.status !== EquipmentAssetStatus.AVAILABLE) {
      throw new BadRequestException('Only available assets can be assigned');
    }

    const employee = await this.employeeRepository.findOne({
      where: { id: dto.employeeId },
    });
    if (!employee) {
      throw new NotFoundException('Employee not found');
    }

    const assignedBy = await this.employeeRepository.findOne({
      where: { id: user.id },
    });

    const now = new Date();
    asset.status = EquipmentAssetStatus.IN_USE;
    asset.assignedEmployee = employee;
    asset.assignedAt = now;
    asset.expectedReturnDate = dto.expectedReturnDate;
    await this.assetRepository.save(asset);

    let request: EquipmentRequest | undefined;
    if (dto.requestId) {
      request =
        (await this.requestRepository.findOne({
          where: { id: dto.requestId },
          relations: { requester: true, equipmentModel: true, category: true },
        })) ?? undefined;
    }

    const assignment = this.assignmentRepository.create({
      equipmentAsset: asset,
      employee,
      request,
      assignedBy: assignedBy ?? undefined,
      assignedAt: now,
      expectedReturnDate: dto.expectedReturnDate,
      status: EquipmentAssignmentStatus.ACTIVE,
    });

    const savedAssignment = await this.assignmentRepository.save(assignment);

    if (request && request.status === RequestStatus.PROCUREMENT_APPROVED) {
      await this.requestRepository.update(request.id, {
        status: RequestStatus.FULFILLED,
      });
      await this.notificationService.notifyEquipmentAssigned(
        request.requester,
        request,
        savedAssignment,
      );
    }

    return savedAssignment;
  }

  async retire(id: string, user: AuthenticatedUser): Promise<EquipmentAsset> {
    const asset = await this.findOne(id);
    if (asset.status === EquipmentAssetStatus.RETIRED) {
      return asset;
    }

    const retiredBy = await this.employeeRepository.findOne({
      where: { id: user.id },
    });

    asset.status = EquipmentAssetStatus.RETIRED;
    asset.retiredAt = new Date();
    asset.retiredBy = retiredBy ?? undefined;
    asset.assignedEmployee = undefined;
    asset.assignedAt = undefined;
    asset.expectedReturnDate = undefined;
    return this.assetRepository.save(asset);
  }

  async remove(id: string): Promise<void> {
    const asset = await this.findOne(id);

    if (asset.status === EquipmentAssetStatus.RETIRED) {
      const retiredAt = asset.retiredAt ? new Date(asset.retiredAt) : null;
      if (retiredAt) {
        const graceEnd = new Date(retiredAt);
        graceEnd.setDate(graceEnd.getDate() + RETIRE_GRACE_PERIOD_DAYS);
        if (new Date() < graceEnd) {
          throw new BadRequestException(
            `Retired assets can only be hard-deleted after ${RETIRE_GRACE_PERIOD_DAYS} days`,
          );
        }
      }
    } else if (await this.hasHistory(asset.id)) {
      throw new BadRequestException(
        'Assets with assignment, request, or procurement history must be retired instead of deleted',
      );
    } else if (
      asset.status !== EquipmentAssetStatus.AVAILABLE &&
      asset.status !== EquipmentAssetStatus.MAINTENANCE
    ) {
      throw new BadRequestException(
        'Only unused available or maintenance assets without history can be hard-deleted',
      );
    }

    await this.assetRepository.remove(asset);
  }

  async hasHistory(assetId: string): Promise<boolean> {
    const assignmentCount = await this.assignmentRepository.count({
      where: { equipmentAsset: { id: assetId } },
    });
    return assignmentCount > 0;
  }

  async getInventoryStats() {
    const assets = await this.assetRepository.find({
      relations: { equipmentModel: true },
    });

    const byStatus = assets.reduce<Record<string, number>>((acc, asset) => {
      acc[asset.status] = (acc[asset.status] ?? 0) + 1;
      return acc;
    }, {});

    const models = await this.modelRepository.find();
    const lowStock = models.filter((model) => {
      const availableCount = assets.filter(
        (asset) =>
          asset.equipmentModel.id === model.id &&
          asset.status === EquipmentAssetStatus.AVAILABLE,
      ).length;
      return availableCount <= model.lowStockThreshold;
    });

    return {
      totalAssets: assets.length,
      byStatus,
      lowStockModels: lowStock.map((model) => ({
        id: model.id,
        name: model.name,
        availableCount: assets.filter(
          (asset) =>
            asset.equipmentModel.id === model.id &&
            asset.status === EquipmentAssetStatus.AVAILABLE,
        ).length,
        lowStockThreshold: model.lowStockThreshold,
      })),
    };
  }

  assertProcurementAccess(user: AuthenticatedUser): void {
    if (!this.accessControl.isProcurementManagerOrAbove(user)) {
      throw new ForbiddenException('Procurement manager access required');
    }
  }
}

File: src/EquipmentApi/Services/EquipmentAssetService.cs
using EquipmentApi.Auth;
using EquipmentApi.Constants;
using EquipmentApi.Data;
using EquipmentApi.Dtos;
using EquipmentApi.Entities;
using EquipmentApi.Enums;
using EquipmentApi.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace EquipmentApi.Services;

public class EquipmentAssetService
{
    private readonly AppDbContext _db;
    private readonly NotificationService _notificationService;
    private readonly AccessControlService _accessControl;

    public EquipmentAssetService(
        AppDbContext db,
        NotificationService notificationService,
        AccessControlService accessControl)
    {
        _db = db;
        _notificationService = notificationService;
        _accessControl = accessControl;
    }

    public Task<List<EquipmentAsset>> FindAllAsync(CancellationToken cancellationToken = default)
    {
        return _db.EquipmentAssets
            .Include(asset => asset.EquipmentModel)
                .ThenInclude(model => model.Category)
            .Include(asset => asset.AssignedEmployee)
            .Include(asset => asset.RetiredBy)
            .OrderByDescending(asset => asset.CreatedAt)
            .ToListAsync(cancellationToken);
    }

    public async Task<EquipmentAsset> FindOneAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var asset = await _db.EquipmentAssets
            .Include(a => a.EquipmentModel)
                .ThenInclude(model => model.Category)
            .Include(a => a.AssignedEmployee)
            .Include(a => a.RetiredBy)
            .Include(a => a.Assignments)
                .ThenInclude(assignment => assignment.Employee)
            .FirstOrDefaultAsync(a => a.Id == id, cancellationToken);

        if (asset is null)
        {
            throw new NotFoundException($"Equipment asset with id \"{id}\" not found");
        }

        return asset;
    }

    public async Task<EquipmentAsset> CreateAsync(
        CreateEquipmentAssetDto dto,
        CancellationToken cancellationToken = default)
    {
        var model = await LoadEquipmentModelAsync(dto.EquipmentModelId, cancellationToken);
        await AssertAssetTagAvailableAsync(dto.AssetTag, cancellationToken);

        var asset = new EquipmentAsset
        {
            EquipmentModelId = model.Id,
            AssetTag = dto.AssetTag,
            SerialNumber = dto.SerialNumber,
            Status = dto.Status ?? EquipmentAssetStatus.Available,
            Notes = dto.Notes,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };

        _db.EquipmentAssets.Add(asset);
        await _db.SaveChangesAsync(cancellationToken);
        asset.EquipmentModel = model;
        return asset;
    }

    public async Task<EquipmentAsset> UpdateAsync(
        Guid id,
        UpdateEquipmentAssetDto dto,
        CancellationToken cancellationToken = default)
    {
        var asset = await FindOneAsync(id, cancellationToken);

        if (dto.AssetTag is not null)
        {
            asset.AssetTag = dto.AssetTag;
        }

        if (dto.SerialNumber is not null)
        {
            asset.SerialNumber = dto.SerialNumber;
        }

        if (dto.Notes is not null)
        {
            asset.Notes = dto.Notes;
        }

        await _db.SaveChangesAsync(cancellationToken);
        return asset;
    }

    public async Task<EquipmentAsset> UpdateStatusAsync(
        Guid id,
        UpdateEquipmentAssetStatusDto dto,
        CancellationToken cancellationToken = default)
    {
        var asset = await FindOneAsync(id, cancellationToken);
        asset.Status = dto.Status;
        await _db.SaveChangesAsync(cancellationToken);
        return asset;
    }

    public async Task<EquipmentAssignment> AssignAsync(
        Guid id,
        AssignEquipmentAssetDto dto,
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var asset = await FindOneAsync(id, cancellationToken);
        if (asset.Status != EquipmentAssetStatus.Available)
        {
            throw new BadRequestException("Only available assets can be assigned");
        }

        var employee = await _db.Employees
            .FirstOrDefaultAsync(e => e.Id == dto.EmployeeId, cancellationToken);
        if (employee is null)
        {
            throw new NotFoundException("Employee not found");
        }

        var now = DateTime.UtcNow;
        asset.Status = EquipmentAssetStatus.InUse;
        asset.AssignedEmployeeId = employee.Id;
        asset.AssignedAt = now;
        asset.ExpectedReturnDate = string.IsNullOrWhiteSpace(dto.ExpectedReturnDate)
            ? null
            : DateOnly.Parse(dto.ExpectedReturnDate);

        EquipmentRequest? request = null;
        if (dto.RequestId is not null)
        {
            request = await _db.EquipmentRequests
                .Include(r => r.Requester)
                .Include(r => r.EquipmentModel)
                .Include(r => r.Category)
                .FirstOrDefaultAsync(r => r.Id == dto.RequestId, cancellationToken);
        }

        var assignment = new EquipmentAssignment
        {
            EquipmentAssetId = asset.Id,
            EmployeeId = employee.Id,
            RequestId = request?.Id,
            AssignedById = user.Id,
            AssignedAt = now,
            ExpectedReturnDate = asset.ExpectedReturnDate,
            Status = EquipmentAssignmentStatus.Active,
            CreatedAt = now,
            UpdatedAt = now,
        };

        _db.EquipmentAssignments.Add(assignment);
        await _db.SaveChangesAsync(cancellationToken);

        if (request is not null && request.Status == RequestStatus.ProcurementApproved)
        {
            request.Status = RequestStatus.Fulfilled;
            await _db.SaveChangesAsync(cancellationToken);
            await _notificationService.NotifyEquipmentAssignedAsync(
                request.Requester,
                request,
                assignment,
                cancellationToken);
        }

        assignment.EquipmentAsset = asset;
        assignment.Employee = employee;
        assignment.Request = request;
        return assignment;
    }

    public async Task<EquipmentAsset> RetireAsync(
        Guid id,
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var asset = await FindOneAsync(id, cancellationToken);
        if (asset.Status == EquipmentAssetStatus.Retired)
        {
            return asset;
        }

        asset.Status = EquipmentAssetStatus.Retired;
        asset.RetiredAt = DateTime.UtcNow;
        asset.RetiredById = user.Id;
        asset.AssignedEmployeeId = null;
        asset.AssignedAt = null;
        asset.ExpectedReturnDate = null;

        await _db.SaveChangesAsync(cancellationToken);
        return asset;
    }

    public async Task RemoveAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var asset = await FindOneAsync(id, cancellationToken);

        if (asset.Status == EquipmentAssetStatus.Retired)
        {
            if (asset.RetiredAt is not null)
            {
                var graceEnd = asset.RetiredAt.Value.AddDays(WorkflowConstants.RetireGracePeriodDays);
                if (DateTime.UtcNow < graceEnd)
                {
                    throw new BadRequestException(
                        $"Retired assets can only be hard-deleted after {WorkflowConstants.RetireGracePeriodDays} days");
                }
            }
        }
        else if (await HasHistoryAsync(asset.Id, cancellationToken))
        {
            throw new BadRequestException(
                "Assets with assignment, request, or procurement history must be retired instead of deleted");
        }
        else if (asset.Status is not EquipmentAssetStatus.Available and not EquipmentAssetStatus.Maintenance)
        {
            throw new BadRequestException(
                "Only unused available or maintenance assets without history can be hard-deleted");
        }

        _db.EquipmentAssets.Remove(asset);
        await _db.SaveChangesAsync(cancellationToken);
    }

    public Task<bool> HasHistoryAsync(Guid assetId, CancellationToken cancellationToken = default)
    {
        return _db.EquipmentAssignments.AnyAsync(
            assignment => assignment.EquipmentAssetId == assetId,
            cancellationToken);
    }

    public async Task<object> GetInventoryStatsAsync(CancellationToken cancellationToken = default)
    {
        var assets = await _db.EquipmentAssets
            .Include(asset => asset.EquipmentModel)
            .ToListAsync(cancellationToken);

        var byStatus = assets
            .GroupBy(asset => asset.Status)
            .ToDictionary(group => group.Key, group => group.Count());

        var models = await _db.EquipmentModels.ToListAsync(cancellationToken);
        var lowStock = models
            .Select(model => new
            {
                model.Id,
                model.Name,
                AvailableCount = assets.Count(asset =>
                    asset.EquipmentModelId == model.Id
                    && asset.Status == EquipmentAssetStatus.Available),
                model.LowStockThreshold,
            })
            .Where(model => model.AvailableCount <= model.LowStockThreshold)
            .ToList();

        return new
        {
            totalAssets = assets.Count,
            byStatus,
            lowStockModels = lowStock,
        };
    }

    public void AssertProcurementAccess(AuthenticatedUser user)
    {
        if (!_accessControl.IsProcurementManagerOrAbove(user))
        {
            throw new ForbiddenException("Procurement manager access required");
        }
    }

    private async Task<EquipmentModel> LoadEquipmentModelAsync(Guid id, CancellationToken cancellationToken)
    {
        var model = await _db.EquipmentModels.FirstOrDefaultAsync(m => m.Id == id, cancellationToken);
        if (model is null)
        {
            throw new NotFoundException("Equipment model not found");
        }

        return model;
    }

    private async Task AssertAssetTagAvailableAsync(string assetTag, CancellationToken cancellationToken)
    {
        var existing = await _db.EquipmentAssets.AnyAsync(asset => asset.AssetTag == assetTag, cancellationToken);
        if (existing)
        {
            throw new ConflictException("Asset tag already exists");
        }
    }
}
